src/Input/Parser.php

```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
     }
 
     /**
-     * Parse single arg.
+     * Parse a single arg.
      *
-     * @param string $arg
+     * @param Token $arg The current token
+     * @param Tokenizer $queue The queue of tokens to be consumed
      *
-     * @return mixed
+     * @return void
      */
-    protected function parseArgs(string $arg)
+    protected function parseArgs(Token $arg, Tokenizer $queue) : void
     {
-        if ($this->_lastVariadic) {
-            return $this->set($this->_lastVariadic, $arg, true);
+        // Handle this argument:
+        $argument = array_shift($this->_arguments);
+        
+        //No argument defined, so its an indexed arg:
+        if (is_null($argument)) {
+
+            //Its a single constant value arg:
+            if ($arg->isConstant()) {
+
+                $this->set(null, $arg->value());
+
+            } else {
+                //Its a variadic arg, so we need to collect all the remaining args:
+                foreach ($arg->nested as $token) {
+                    if ($token->isConstant()) {
+                        $this->set(null, $token->value(), true);
+                    } else {
+                        throw new InvalidParameterException("Only constant parameters are allowed in variadic arguments");
+                    }
+                }
+            }
+            return;
         }
 
-        if (!$argument = reset($this->_arguments)) {
-            return $this->set(null, $arg);
+        // Its variadic, so we need to collect all the remaining args:
+        if ($argument->variadic() && $arg->isConstant()) {
+
+            // Consume all the remaining tokens
```

Comment:
seems the comment is short enough to fit in single line:
`// Consume all the remaining tokens if an option is found, treat it as well`